lib/web.ts

```diff
@@ -219,24 +218,63 @@ export default class AutoConsent {
     return foundCMPs;
   }
 
-  async detectPopups(cmps: AutoCMP[]): Promise<AutoCMP[]> {
-    const result: AutoCMP[] = [];
-    const popupLookups = cmps.map((cmp) => this.waitForPopup(cmp).then((isOpen) => {
-      if (isOpen) {
-        this.updateState({ detectedPopups: this.state.detectedPopups.concat([cmp.name]) });
-        this.sendContentMessage({
-          type: 'popupFound',
-          cmp: cmp.name,
-          url: location.href,
-        }); // notify the browser
-        result.push(cmp);
+  async detectPopup(cmp: AutoCMP): Promise<AutoCMP> {
+    const isOpen = await this.waitForPopup(cmp);
+
+    if (isOpen) {
+      this.updateState({ detectedPopups: this.state.detectedPopups.concat([cmp.name]) });
+      this.sendContentMessage({
+        type: 'popupFound',
+        cmp: cmp.name,
+        url: location.href,
+      }); // notify the browser
+    }
+
+    return cmp
+  }
+
+  detectPopups(cmps: AutoCMP[], onFirstPopupAppears: (cmp: AutoCMP) => unknown): Promise<AutoCMP[]> {
+    return new Promise(resolve => {
+      const foundCmps: AutoCMP[] = []
+      let completed = 0
+
+      for (let i = 0, l = cmps.length; i < l; i++) {
+        this.detectPopup(cmps[i])
+          .then(cmp => {
+            foundCmps.push(cmp)
+            
+            if (foundCmps.length === 1) {
+              onFirstPopupAppears(cmp)
+            }
+          })
+          .catch(error => {
+            enableLogs && console.warn(`error waiting for a popup for ${cmps[i].name}`, error)
+          })
+          .finally(() => {
+            if (++completed === l) {
+              resolve(foundCmps)
+            }
```

Comment:
Could this be rewritten with `async`/`await`? I find that easier to follow, and matches the code-style in the rest of the project.

Reply:
I converted the function using `Promise.race` and `Promise.allSettled`. Generally agreeing to code-style problem.

Reply:
* changing as later review stated that Promise.any should be used.